src/Shared/Metrics/MetricsExtensions.cs

```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Http;
+
+internal static class MetricsExtensions
+{
+    public static bool TryAddTag(this IHttpMetricsTagsFeature feature, string name, object? value)
+    {
+        var tags = feature.Tags;
+
+        // Tags is internally represented as a List<T>.
+        // Prefer looping through the list to avoid allocating an enumerator.
+        if (tags is List<KeyValuePair<string, object?>> list)
+        {
+            foreach (var tag in list)
```

Comment:
It's been a while, but I think foreach might do this for all `IList`s.

Reply:
I'm not sure what you mean. The type on the feature is `ICollection<T>` (IMO we should have made it `IList<T>` but took late now). A foreach will always allocate an enumerator, no?

Reply:
I misremembered - there's special handling for arrays and strings, but not for ILists.